llvm/lib/Target/RISCV/MCTargetDesc/RISCVMCCodeEmitter.cpp

```diff
@@ -467,11 +467,15 @@ unsigned RISCVMCCodeEmitter::getImmOpValue(const MCInst &MI, unsigned OpNo,
       FixupKind = RISCV::fixup_riscv_rvc_jump;
     } else if (MIFrm == RISCVII::InstFormatCB) {
       FixupKind = RISCV::fixup_riscv_rvc_branch;
-    } else if (MIFrm == RISCVII::InstFormatCVHWLP) {
-      if (OpNo == 1 && MI.getNumOperands() == 3) {
-        // This is the cv.setupi instruction and therefore requires the 5 bit
-        // relocation.
-        FixupKind = RISCV::fixup_riscv_cvpcrel_urs1;
+    } else if (MIFrm == RISCVII::InstFormatCVHwlp) {
+      if (MI.getNumOperands() == 3) {
+        auto &MOCount = MI.getOperand(1);
+        if (MOCount.isReg())
+          FixupKind = RISCV::fixup_riscv_cvpcrel_ui12;
+        else
+          // This is the cv.setupi instruction and therefore requires the 5 bit
+          // relocation.
+          FixupKind = RISCV::fixup_riscv_cvpcrel_urs1;
```

Comment:
This change is because of the internal swap of the operands for `cv.setupi`. It checks the number of operands (only the `cv.setup[i]` instructions have 3 operands) and then checks if the second operand is a register operand to distinguish the `setup` from the `setupi` instruction.

This also makes this check more robust.